RecoTracker/LSTCore/src/alpaka/Quintuplet.h

```diff
@@ -1712,107 +1722,158 @@ namespace ALPAKA_ACCELERATOR_NAMESPACE::lst {
-
-            float t5Embed[Params_T5::kEmbed] = {0.f};
-
-            bool tightCutFlag = false;
-            bool success = runQuintupletDefaultAlgo(acc,
-                                                    modules,
-                                                    mds,
-                                                    segments,
-                                                    triplets,
-                                                    lowerModule1,
-                                                    lowerModule2,
-                                                    lowerModule3,
-                                                    lowerModule4,
-                                                    lowerModule5,
-                                                    innerTripletIndex,
-                                                    outerTripletIndex,
-                                                    innerRadius,
-                                                    outerRadius,
-                                                    bridgeRadius,
-                                                    regressionCenterX,
-                                                    regressionCenterY,
-                                                    regressionRadius,
-                                                    rzChiSquared,
-                                                    chiSquared,
-                                                    nonAnchorChiSquared,
-                                                    dBeta1,
-                                                    dBeta2,
-                                                    tightCutFlag,
-                                                    t5Embed,
-                                                    ptCut);
-
-            if (success) {
-              int totOccupancyQuintuplets = alpaka::atomicAdd(
-                  acc, &quintupletsOccupancy.totOccupancyQuintuplets()[lowerModule1], 1u, alpaka::hierarchy::Threads{});
-              if (totOccupancyQuintuplets >= ranges.quintupletModuleOccupancy()[lowerModule1]) {
+
+            unsigned int secondSegmentIndex = triplets.segmentIndices()[innerTripletIndex][1];
+            unsigned int thirdSegmentIndex = triplets.segmentIndices()[outerTripletIndex][0];
+
+            unsigned int miniDoublet3Index =
+                segments.mdIndices()[secondSegmentIndex][1];  //inner triplet outer segment outer MD index
+            unsigned int outerInnerInnerMiniDoubletIndex =
+                segments.mdIndices()[thirdSegmentIndex][0];  //outer triplet inner segment inner MD index
+
+            //this cut reduces the number of candidates by a factor of 3, i.e., 2 out of 3 warps can end right here!
+            if (miniDoublet3Index != outerInnerInnerMiniDoubletIndex)
+              continue;
+
+            // Match inner Sg and Outer Sg
+            int mIdx = alpaka::atomicAdd(acc, &matchCount, 1, alpaka::hierarchy::Blocks{});
```

Comment:
`matchCount` is a block-level shared variable, so the atomic operations can be restricted to the block scope:
```suggestion
            int mIdx = alpaka::atomicAdd(acc, &matchCount, 1, alpaka::hierarchy::Threads{});
```

(`Threads` means "all the threads in a block", while `Blocks` means "all the blocks in a grid" - and yes, I find it confusing too)

Reply:
> Can you point us to a documentation explaining the hierarchies?

Nope, sorry 🤷🏻‍♂️ 

The atomic operations are:
  - `alpaka::hierarchy::Threads{}`: atomic across all __thread__ in a block;
  - `alpaka::hierarchy::Blocks{}`: atomic across all __blocks__ in a grid;
  - `alpaka::hierarchy::Grids{}`: atomic for all __grids__ (_i.e._ kernels) on the device.

Again, I agree that the naming is confusing, and the documentation is lacking.
If you would open an issue about it on the alpaka GitHub page https://github.com/alpaka-group/alpaka/issues/ I will follow up with the other developers.

On the other hand, there are already 20 open issues regarding atomics...